server/pulp/plugins/conduits/mixins.py

```diff
@@ -699,15 +700,17 @@ def do_get_repo_units(repo_id, criteria, exception_class, as_generator=False):
         # Use a get_units as_generator here and cast to a list later, if necessary.
         units = association_query_manager.get_units(repo_id, criteria=criteria, as_generator=True)
 
-        # Load all type definitions so we don't hammer the database.
-        type_defs = dict((t['id'], t) for t in types_db.all_type_definitions())
-
         # Transfer object generator.
         def _transfer_object_generator():
+            unit_key_fields_cache = {}
             for u in units:
                 type_id = u['unit_type_id']
-                type_def = type_defs[type_id]
-                yield common_utils.to_plugin_associated_unit(u, type_id, type_def['unit_key'])
+                if type_id not in unit_key_fields_cache:
```

Comment:
Maybe use setdefault here? I recently learned about it [from this site](http://docs.quantifiedcode.com/python-anti-patterns/correctness/not_using_setdefault_to_initialize_a_dictionary.html).